core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,395 @@
+      @Override
+      public void addListener(KeyAwareLockListener listener) {
+         lockPromise.addListener(state -> listener.onEvent(key, state));
+      }
+
+      @Override
+      public void run() {
+         lockPromise.cancel(LockState.TIMED_OUT);
+      }
+
+      public KeyAwareExtendedLockPromise scheduleLockTimeoutTask(ScheduledExecutorService executorService) {
+         if (executorService != null && timeoutMillis > 0 && !isAvailable()) {
+            ScheduledFuture<?> future = executorService.schedule(this, timeoutMillis, TimeUnit.MILLISECONDS);
+            lockPromise.addListener((state -> future.cancel(false)));
+         }
+         return this;
+      }
+   }
+
+   private static class CompositeLockPromise implements KeyAwareLockPromise, LockListener, Runnable {
+
+      private final List<KeyAwareExtendedLockPromise> lockPromiseList;
+      private final CompletableFuture<Void> notifier;
+      volatile LockState lockState = LockState.AVAILABLE;
+
+      private CompositeLockPromise(int size) {
+         lockPromiseList = new ArrayList<>(size);
+         notifier = new CompletableFuture<>();
+      }
+
+      public void addLock(KeyAwareExtendedLockPromise lockPromise) {
+         lockPromiseList.add(lockPromise);
+      }
+
+      public void markListAsFinal() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            lockPromise.addListener(this);
+         }
+      }
+
+      @Override
+      public boolean isAvailable() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            if (!lockPromise.isAvailable()) {
+               return false;
+            }
+         }
+         return true;
+      }
+
+      @Override
+      public void lock() throws InterruptedException, TimeoutException {
+         Exception exception = null;
+         ExceptionType exceptionType = ExceptionType.NONE;
+         for (ExtendedLockPromise lockPromise : lockPromiseList) {
+            try {
+               //we still need to invoke lock in all the locks.
+               lockPromise.lock();
+            } catch (InterruptedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.INTERRUPTED;
+               }
+            } catch (TimeoutException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.TIMEOUT;
+               }
+            } catch (DeadlockDetectedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.DEADLOCK;
+               }
+            } catch (RuntimeException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.RUNTIME;
+               }
+            }
+         }
+         if (exception != null) {
+            switch (exceptionType) {
+               case INTERRUPTED:
```

Comment:
I think `exception instanceof InterruptedException` should also work... But I think maybe we want some kind of exception hierarchy, e.g. `InterruptedException` trumps everything else, in which case a separate variable for each exception type would be better.

Also, I think we should try to preserve all the exceptions when there are more of them, maybe with `addSuppressed()`.